### PR TITLE
SAP: Improve stopping criteria

### PR DESCRIPTION
**Pull request by amcastro-tri**
Given we merged #16240 quickly due to its high priority, I address here the review comments from #16240.

I like this version significantly better since we can get high accuracy while at the same time we can exit gracefully if the problem does not make any progress due to round-off errors (that is, the solver did indeed reach the best solution modulo round-off errors).

---
This change is [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/16243)

### CODE REVIEW

**Review by amcastro-tri — COMMENTED** [review]
+@xuchenhan-tri for feature review please.

@sherm1, PTAL to see if I addressed your question from #16240. 

> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/16243#-:-Mr8fIxY2YJboG7VSvjk:b-whl70a)* status: LGTM missing from assignee xuchenhan-tri, needs platform reviewer assigned, needs at least two assigned reviewers (waiting on @xuchenhan-tri)

**Comment by xuchenhan-tri** [comment]
@drake-jenkins-bot mac-big-sur-clang-bazel-continuous-release please.

**Comment by xuchenhan-tri** [comment]
*[a discussion](https://reviewable.io/reviews/robotlocomotion/drake/16243#-Mr8kNJk8wR-HBUEg8wq:-Mr8kNJk8wR-HBUEg8wr:bo7t79j) (no related file):*
@drake-jenkins-bot mac-big-sur-clang-bazel-continuous-release please.

---

**Comment by amcastro-tri** [comment]
fyi @xuchenhan-tri, I did test this in macsim, so unless there is a new surprise, I'd expect this to pass the mac tests. (I was hoping to burn CI hours after the review, but I can see how this would buy you piece of mind now)

**Review by xuchenhan-tri — COMMENTED** [review]
> Reviewed 1 of 3 files at r1, 1 of 1 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/16243)* status: 4 unresolved discussions, LGTM missing from assignee xuchenhan-tri, needs platform reviewer assigned, needs at least two assigned reviewers (waiting on @amcastro-tri and @xuchenhan-tri)

___
*[multibody/contact_solvers/sap_solver.h, line 26 at r2](https://reviewable.io/reviews/robotlocomotion/drake/16243#-Mr8njfY9dsi7UjCr_S7:-Mr8njfY9dsi7UjCr_S8:bdsjv67) ([raw file](https://github.com/robotlocomotion/drake/blob/4421e18e6c2f6ae345f8d86b95140c3cb705de9d/multibody/contact_solvers/sap_solver.h#L26)):*
> ```objc
>   // momentum equally.
>   //
>   // Nominal values:
> ```

I don't understand where these nominal values come from. It seems to me that whether the optimality condition reaches or the cost condition reaches is a function of three things

1. Tolerance for optimality condition
2. Tolerance for cost condition
3. Conditioning of the problem.

The guideline here discusses 1. independent of 2, 3. That doesn't make sense to me.
___
*[multibody/contact_solvers/sap_solver.h, line 58 at r2](https://reviewable.io/reviews/robotlocomotion/drake/16243#-Mr8oSR0B77rtd9Cj031:-Mr8oSR10i_6CalMSkDi:b-iqyb1n) ([raw file](https://github.com/robotlocomotion/drake/blob/4421e18e6c2f6ae345f8d86b95140c3cb705de9d/multibody/contact_solvers/sap_solver.h#L58)):*
> ```objc
>   // SolverStats::cost_condition_reached indicates if this condition was
>   // reached.
>   double cost_abs_tolerance{1.e-30};  // Absolute tolerance εₐ, in Joules.
> ```

Looks like these values are chosen according to floating point precision? If that's the case, I think they should be a constant parameter instead of something we give the users to tune.
___
*[multibody/contact_solvers/sap_solver.cc, line 502 at r2](https://reviewable.io/reviews/robotlocomotion/drake/16243#-Mr8q8By1LcksYShwqrj:-Mr8q8By1LcksYShwqrk:b91w2a0) ([raw file](https://github.com/robotlocomotion/drake/blob/4421e18e6c2f6ae345f8d86b95140c3cb705de9d/multibody/contact_solvers/sap_solver.cc#L502)):*
> ```cpp
>   // is zero.
>   const T ell_scale = 0.5 * (ell + ell0);
>   if (abs(dell_dalpha0 / ell_scale) < kTolerance) return std::make_pair(1.0, 0);
> ```

Could you explain why this exit condition?

Looks like we are giving up on the line search here because dell_dalpha0 is too small. What does that say about convergence? Does that guarantee we exit via the cost condition in the newton iterations?

**Review by xuchenhan-tri — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/16243)* status: 5 unresolved discussions, LGTM missing from assignee xuchenhan-tri, needs platform reviewer assigned, needs at least two assigned reviewers (waiting on @amcastro-tri and @xuchenhan-tri)

___
*[multibody/contact_solvers/test/sap_solver_test.cc, line 454 at r2](https://reviewable.io/reviews/robotlocomotion/drake/16243#-Mr907rJ1-HyVbJ_eKNF:-Mr907rJ1-HyVbJ_eKNG:b-zal5oe) ([raw file](https://github.com/robotlocomotion/drake/blob/4421e18e6c2f6ae345f8d86b95140c3cb705de9d/multibody/contact_solvers/test/sap_solver_test.cc#L454)):*
> ```cpp
>   sap.set_parameters(params);
>   ContactSolverResults<double> result;
>   // Arbitrary non-zero guess to stress the solver.
> ```

BTW, does the initial guess now _have to_ be nonzero in order to trigger cost condition when that's desired?
___
*[multibody/contact_solvers/test/sap_solver_test.cc, line 660 at r2](https://reviewable.io/reviews/robotlocomotion/drake/16243#-Mr9-NVHAJWvQSZY10IM:-Mr9-NVHAJWvQSZY10IN:b-5w1cyy) ([raw file](https://github.com/robotlocomotion/drake/blob/4421e18e6c2f6ae345f8d86b95140c3cb705de9d/multibody/contact_solvers/test/sap_solver_test.cc#L660)):*
> ```cpp
> }
> 
> // We set a very tight optimality tolerance. The solver wont be able to reach
> ```

nit, s/wont/won't
___
*[multibody/contact_solvers/test/sap_solver_test.cc, line 672 at r2](https://reviewable.io/reviews/robotlocomotion/drake/16243#-Mr9-HqT7jg19R-ntqnd:-Mr9-HqT7jg19R-ntqne:b-d1hojd) ([raw file](https://github.com/robotlocomotion/drake/blob/4421e18e6c2f6ae345f8d86b95140c3cb705de9d/multibody/contact_solvers/test/sap_solver_test.cc#L672)):*
> ```cpp
> GTEST_TEST(PizzaSaver, StictionAtTightOptimalityTolerance) {
>   SapSolverParameters params;
>   // Extremely tight tolerances to trigger the cost stoping criterion.
> ```

nit, s/stoping/stopping

**Review by xuchenhan-tri — COMMENTED** [review]
> Reviewed 1 of 3 files at r1.
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/16243)* status: 5 unresolved discussions, LGTM missing from assignee xuchenhan-tri, needs platform reviewer assigned, needs at least two assigned reviewers (waiting on @amcastro-tri)

**Review by amcastro-tri — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/16243)* status: 3 unresolved discussions, LGTM missing from assignee xuchenhan-tri, needs platform reviewer assigned, needs at least two assigned reviewers (waiting on @xuchenhan-tri)

___
*[multibody/contact_solvers/sap_solver.h, line 26 at r2](https://reviewable.io/reviews/robotlocomotion/drake/16243#-Mr8njfY9dsi7UjCr_S7:-Mr9-qcmAyh8Je3LbnyY:bna6ca2) ([raw file](https://github.com/robotlocomotion/drake/blob/4421e18e6c2f6ae345f8d86b95140c3cb705de9d/multibody/contact_solvers/sap_solver.h#L26)):*
> I don't understand where these nominal values come from

From experience. Essentially the values used and reported in our paper.

> independent of 2, 3. That doesn't make sense to me.

I am not sure I follow your concern, but I believe we can decouple these conditions. Which one we reach first is however a function of those 3 bullet points. I was hoping the documentation of the cost condition below would fill this gap. What do you think it's missing?
___
*[multibody/contact_solvers/sap_solver.h, line 58 at r2](https://reviewable.io/reviews/robotlocomotion/drake/16243#-Mr8oSR0B77rtd9Cj031:-Mr90dw4D3L04-aZgKii:bh7rq62) ([raw file](https://github.com/robotlocomotion/drake/blob/4421e18e6c2f6ae345f8d86b95140c3cb705de9d/multibody/contact_solvers/sap_solver.h#L58)):*
<details><summary><i>Previously, xuchenhan-tri wrote…</i></summary><blockquote>

Looks like these values are chosen according to floating point precision? If that's the case, I think they should be a constant parameter instead of something we give the users to tune.
</blockquote></details>

I see your point. We might set them on stone once we are happy with them after lots of experimentation (a year?). For now I'd prefer having a quick option to set them. Also, don't forget that the user here is DiscreteUpdateManager, which in turn will most likely expose a much limited number of parameters to the final user (a person), if any.
___
*[multibody/contact_solvers/sap_solver.cc, line 502 at r2](https://reviewable.io/reviews/robotlocomotion/drake/16243#-Mr8q8By1LcksYShwqrj:-Mr91J52B_Q22xw7r5xz:b-62pz69) ([raw file](https://github.com/robotlocomotion/drake/blob/4421e18e6c2f6ae345f8d86b95140c3cb705de9d/multibody/contact_solvers/sap_solver.cc#L502)):*
<details><summary><i>Previously, xuchenhan-tri wrote…</i></summary><blockquote>

Could you explain why this exit condition?

Looks like we are giving up on the line search here because dell_dalpha0 is too small. What does that say about convergence? Does that guarantee we exit via the cost condition in the newton iterations? 
</blockquote></details>

We are not giving up, but detecting that we are close to the solution and line search is not necessary (nor well defined). I added comments, PTAL.
___
*[multibody/contact_solvers/test/sap_solver_test.cc, line 454 at r2](https://reviewable.io/reviews/robotlocomotion/drake/16243#-Mr907rJ1-HyVbJ_eKNF:-Mr92JoM3glKplCb9nSq:b-hli9b4) ([raw file](https://github.com/robotlocomotion/drake/blob/4421e18e6c2f6ae345f8d86b95140c3cb705de9d/multibody/contact_solvers/test/sap_solver_test.cc#L454)):*
<details><summary><i>Previously, xuchenhan-tri wrote…</i></summary><blockquote>

BTW, does the initial guess now _have to_ be nonzero in order to trigger cost condition when that's desired?
</blockquote></details>

I haven't checked. The reason to take an arbitrary non-zero guess is to stress the solver.

**Review by sherm1 — COMMENTED** [review]
Platform :lgtm: pending a few comments and feature review

> Reviewed 2 of 3 files at r1, 1 of 1 files at r2, 2 of 2 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/16243#-:-Mr96miq-f95kBTptEVk:bh52pmj)* status: 5 unresolved discussions, LGTM missing from assignee xuchenhan-tri, needs platform reviewer assigned, needs at least two assigned reviewers (waiting on @amcastro-tri and @xuchenhan-tri)

___
*[multibody/contact_solvers/sap_solver.h, line 59 at r2](https://reviewable.io/reviews/robotlocomotion/drake/16243#-Mr8jmNaDTUw1p5BR6Wl:-Mr8jmNbDnJTlTmBWZ-k:blm2m67) ([raw file](https://github.com/robotlocomotion/drake/blob/4421e18e6c2f6ae345f8d86b95140c3cb705de9d/multibody/contact_solvers/sap_solver.h#L59)):*
> ```objc
>   // reached.
>   double cost_abs_tolerance{1.e-30};  // Absolute tolerance εₐ, in Joules.
>   double cost_rel_tolerance{1.e-15};  // Relative tolerance εᵣ.
> ```

BTW wow, those are really strict. Can we really work with an absolute tolerance that tight? That means we keep going if the change in cost near zero is, say, 1e-25?
___
*[multibody/contact_solvers/sap_solver.cc, line 401 at r2](https://reviewable.io/reviews/robotlocomotion/drake/16243#-Mr8kOqI31oVO6nSMfER:-Mr8kOqI31oVO6nSMfES:b-zi9gjl) ([raw file](https://github.com/robotlocomotion/drake/blob/4421e18e6c2f6ae345f8d86b95140c3cb705de9d/multibody/contact_solvers/sap_solver.cc#L401)):*
> ```cpp
>     const double ell_decrement = std::abs(ell_previous - ell);
> 
>     // SAP's convergence is monotonous. We sanity check this here. We use a slop
> ```

nit: monotonous -> monotonic
("monotonous" means "boring" :)

**Review by amcastro-tri — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/16243)* status: 3 unresolved discussions, LGTM missing from assignee xuchenhan-tri, needs platform reviewer assigned, needs at least two assigned reviewers (waiting on @sherm1 and @xuchenhan-tri)

___
*[multibody/contact_solvers/sap_solver.h, line 59 at r2](https://reviewable.io/reviews/robotlocomotion/drake/16243#-Mr8jmNaDTUw1p5BR6Wl:-Mr9DW1u6ah5qEPte_jE:b-wxo1b2) ([raw file](https://github.com/robotlocomotion/drake/blob/4421e18e6c2f6ae345f8d86b95140c3cb705de9d/multibody/contact_solvers/sap_solver.h#L59)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

BTW wow, those are really strict. Can we really work with an absolute tolerance that tight? That means we keep going if the change in cost near zero is, say, 1e-25?
</blockquote></details>

I added notes on nominal values.
___
*[multibody/contact_solvers/sap_solver.cc, line 401 at r2](https://reviewable.io/reviews/robotlocomotion/drake/16243#-Mr8kOqI31oVO6nSMfER:-Mr9DbL_EEV-Dd-xi3Bc:b-x499e3) ([raw file](https://github.com/robotlocomotion/drake/blob/4421e18e6c2f6ae345f8d86b95140c3cb705de9d/multibody/contact_solvers/sap_solver.cc#L401)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

nit: monotonous -> monotonic
("monotonous" means "boring" :)
</blockquote></details>

agh.... thanks!

**Review by xuchenhan-tri — COMMENTED** [review]
> Reviewed 2 of 2 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/16243)* status: 3 unresolved discussions, LGTM missing from assignee xuchenhan-tri, needs platform reviewer assigned, needs at least two assigned reviewers (waiting on @amcastro-tri)

___
*[multibody/contact_solvers/sap_solver.h, line 26 at r2](https://reviewable.io/reviews/robotlocomotion/drake/16243#-Mr8njfY9dsi7UjCr_S7:-Mr9E78s8DAOMHP2MQxU:b35nsbj) ([raw file](https://github.com/robotlocomotion/drake/blob/4421e18e6c2f6ae345f8d86b95140c3cb705de9d/multibody/contact_solvers/sap_solver.h#L26)):*
Take the first nominal value for example: 

>   //  * For rel_tolerance > 1.0e-5 the solver will have no problem reaching this condition.
  
Suppose I set the tolerance for cost condition unreasonably large. Then the newton iterations will exit without reaching the 1e-5 tolerance. Is that right? Without discussion about bullet point 2 and 3, I don't see the value of these nominal values.
___
*[multibody/contact_solvers/sap_solver.h, line 58 at r2](https://reviewable.io/reviews/robotlocomotion/drake/16243#-Mr8oSR0B77rtd9Cj031:-Mr9Edp293LC9aNptPMd:bxim3xa) ([raw file](https://github.com/robotlocomotion/drake/blob/4421e18e6c2f6ae345f8d86b95140c3cb705de9d/multibody/contact_solvers/sap_solver.h#L58)):*
>  Also, don't forget that the user here is DiscreteUpdateManager, which in turn will most likely expose a much limited number of parameters to the final user (a person), if any.

Yeah, I missed that point. In that case, I can see the benefit of leaving a knob for just us to tune.
___
*[multibody/contact_solvers/sap_solver.cc, line 502 at r2](https://reviewable.io/reviews/robotlocomotion/drake/16243#-Mr8q8By1LcksYShwqrj:-Mr9FF_TAE4PpETXvryR:bimu7bm) ([raw file](https://github.com/robotlocomotion/drake/blob/4421e18e6c2f6ae345f8d86b95140c3cb705de9d/multibody/contact_solvers/sap_solver.cc#L502)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

We are not giving up, but detecting that we are close to the solution and line search is not necessary (nor well defined). I added comments, PTAL.
</blockquote></details>

Let me rephrase the concern: is it possible that we return alpha = 1 here and end up not converging in the newton iterations?
___
*[multibody/contact_solvers/test/sap_solver_test.cc, line 454 at r2](https://reviewable.io/reviews/robotlocomotion/drake/16243#-Mr907rJ1-HyVbJ_eKNF:-Mr9FRpe-V3ICIJ7pKHP:b-vtgiip) ([raw file](https://github.com/robotlocomotion/drake/blob/4421e18e6c2f6ae345f8d86b95140c3cb705de9d/multibody/contact_solvers/test/sap_solver_test.cc#L454)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

I haven't checked. The reason to take an arbitrary non-zero guess is to stress the solver.
</blockquote></details>

Now that we have two exit conditions and test that insists on a particular exit condition being triggered, `v_guess` being nonzero is not just an extra mile we go to stress the solver, but a necessity for the test to pass. 

In fact, I just tried changing `v_guess` to zero and the unit test failed.

**Review by sherm1 — COMMENTED** [review]
> Reviewed 2 of 2 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/16243)* status: 3 unresolved discussions, LGTM missing from assignee xuchenhan-tri, needs platform reviewer assigned, needs at least two assigned reviewers (waiting on @amcastro-tri)

**Review by amcastro-tri — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/16243)* status: 3 unresolved discussions, LGTM missing from assignee xuchenhan-tri, needs platform reviewer assigned, needs at least two assigned reviewers (waiting on @amcastro-tri and @xuchenhan-tri)

___
*[multibody/contact_solvers/sap_solver.h, line 26 at r2](https://reviewable.io/reviews/robotlocomotion/drake/16243#-Mr8njfY9dsi7UjCr_S7:-Mr9HblaBsspShO2bcY6:b-yjlhfv) ([raw file](https://github.com/robotlocomotion/drake/blob/4421e18e6c2f6ae345f8d86b95140c3cb705de9d/multibody/contact_solvers/sap_solver.h#L26)):*
<details><summary><i>Previously, xuchenhan-tri wrote…</i></summary><blockquote>

Take the first nominal value for example: 

>   //  * For rel_tolerance > 1.0e-5 the solver will have no problem reaching this condition.
  
Suppose I set the tolerance for cost condition unreasonably large. Then the newton iterations will exit without reaching the 1e-5 tolerance. Is that right? Without discussion about bullet point 2 and 3, I don't see the value of these nominal values.

</blockquote></details>

I think you are right. Do you think the information you need is covered in the documentation of the cost tolerances and then making the proper connection is enough? or do you really think there is something else missing?
___
*[multibody/contact_solvers/sap_solver.cc, line 502 at r2](https://reviewable.io/reviews/robotlocomotion/drake/16243#-Mr8q8By1LcksYShwqrj:-Mr9HpUW6QHUnbY81pnM:bf6ajtn) ([raw file](https://github.com/robotlocomotion/drake/blob/4421e18e6c2f6ae345f8d86b95140c3cb705de9d/multibody/contact_solvers/sap_solver.cc#L502)):*
<details><summary><i>Previously, xuchenhan-tri wrote…</i></summary><blockquote>

Let me rephrase the concern: is it possible that we return alpha = 1 here and end up not converging in the newton iterations?
</blockquote></details>

We are in round-off error territory here, so we can forget about Armijo. I think the worst that could happen is that the solver makes no progress any longer, and then the cost condition would be reached.

**Review by xuchenhan-tri — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/16243)* status: 3 unresolved discussions, LGTM missing from assignee xuchenhan-tri, needs platform reviewer assigned, needs at least two assigned reviewers (waiting on @amcastro-tri)

___
*[multibody/contact_solvers/sap_solver.h, line 26 at r2](https://reviewable.io/reviews/robotlocomotion/drake/16243#-Mr8njfY9dsi7UjCr_S7:-Mr9KoucEQI_EBig-6ln:bjib7mj) ([raw file](https://github.com/robotlocomotion/drake/blob/4421e18e6c2f6ae345f8d86b95140c3cb705de9d/multibody/contact_solvers/sap_solver.h#L26)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

I think you are right. Do you think the information you need is covered in the documentation of the cost tolerances and then making the proper connection is enough? or do you really think there is something else missing?
</blockquote></details>

What's missing IMO is the disclaimer that these values only serve as a useful guideline with the default cost tolerance values and typical examples that we have tried so far. Users should use their own discretion instead of treating these values as the golden rules.

Perhaps that's what you meant by nominal values and I missed that? Anyway, I think this point needs to be stressed.
___
*[multibody/contact_solvers/sap_solver.cc, line 502 at r2](https://reviewable.io/reviews/robotlocomotion/drake/16243#-Mr8q8By1LcksYShwqrj:-Mr9LOBg5o0LNCaCGdlD:b-2fu1bz) ([raw file](https://github.com/robotlocomotion/drake/blob/4421e18e6c2f6ae345f8d86b95140c3cb705de9d/multibody/contact_solvers/sap_solver.cc#L502)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

We are in round-off error territory here, so we can forget about Armijo. I think the worst that could happen is that the solver makes no progress any longer, and then the cost condition would be reached.
</blockquote></details>

Since we are forgoing Armijo, there's no guarantee that optimality condition will reach. I can see it's reasonable to believe that cost condition will eventually be triggered and we exit the solve the  "success" status.

My concern is: is the cost condition guaranteed to be triggered in this case? If not, then we would exit the solve with a "failed" status. Not sure if that's what we want.

**Review by amcastro-tri — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/16243)* status: 3 unresolved discussions, LGTM missing from assignee xuchenhan-tri, needs platform reviewer assigned, needs at least two assigned reviewers (waiting on @amcastro-tri and @xuchenhan-tri)

___
*[multibody/contact_solvers/sap_solver.cc, line 502 at r2](https://reviewable.io/reviews/robotlocomotion/drake/16243#-Mr8q8By1LcksYShwqrj:-Mr9XfI77dyYx6EBxSGD:bnncu61) ([raw file](https://github.com/robotlocomotion/drake/blob/4421e18e6c2f6ae345f8d86b95140c3cb705de9d/multibody/contact_solvers/sap_solver.cc#L502)):*
<details><summary><i>Previously, xuchenhan-tri wrote…</i></summary><blockquote>

Since we are forgoing Armijo, there's no guarantee that optimality condition will reach. I can see it's reasonable to believe that cost condition will eventually be triggered and we exit the solve the  "success" status.

My concern is: is the cost condition guaranteed to be triggered in this case? If not, then we would exit the solve with a "failed" status. Not sure if that's what we want.
</blockquote></details>

I think exciting with a fail status is reasonable. If that happens, we can revisit this condition and find out how we can do better. IMO this is an improvement on the original PR. 
Not sure how we can do better.

**Review by xuchenhan-tri — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/16243)* status: 2 unresolved discussions, LGTM missing from assignee xuchenhan-tri, needs platform reviewer assigned, needs at least two assigned reviewers (waiting on @amcastro-tri)

___
*[multibody/contact_solvers/sap_solver.cc, line 502 at r2](https://reviewable.io/reviews/robotlocomotion/drake/16243#-Mr8q8By1LcksYShwqrj:-Mr9ZH4e-WXCNH2msxsY:bgszoky) ([raw file](https://github.com/robotlocomotion/drake/blob/4421e18e6c2f6ae345f8d86b95140c3cb705de9d/multibody/contact_solvers/sap_solver.cc#L502)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

I think exciting with a fail status is reasonable. If that happens, we can revisit this condition and find out how we can do better. IMO this is an improvement on the original PR. 
Not sure how we can do better.
</blockquote></details>

Ok, just want to make sure you have considered this case. Thanks!

**Review by xuchenhan-tri — COMMENTED** [review]
:lgtm: pending a couple documentation requests.

> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/16243#-:-Mr9ZR3s7g8561TIvzwy:boy4efv)* status: 2 unresolved discussions, needs platform reviewer assigned, needs at least two assigned reviewers (waiting on @amcastro-tri)

**Review by amcastro-tri — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/16243)* status: 2 unresolved discussions, needs platform reviewer assigned, needs at least two assigned reviewers (waiting on @amcastro-tri, @sherm1, and @xuchenhan-tri)

___
*[multibody/contact_solvers/sap_solver.h, line 26 at r2](https://reviewable.io/reviews/robotlocomotion/drake/16243#-Mr8njfY9dsi7UjCr_S7:-Mr9Z_9t6wD0fkdnzaWy:b-penjsm) ([raw file](https://github.com/robotlocomotion/drake/blob/4421e18e6c2f6ae345f8d86b95140c3cb705de9d/multibody/contact_solvers/sap_solver.h#L26)):*
<details><summary><i>Previously, xuchenhan-tri wrote…</i></summary><blockquote>

What's missing IMO is the disclaimer that these values only serve as a useful guideline with the default cost tolerance values and typical examples that we have tried so far. Users should use their own discretion instead of treating these values as the golden rules.

Perhaps that's what you meant by nominal values and I missed that? Anyway, I think this point needs to be stressed.
</blockquote></details>

I added more docs. Hopefully that clarifies? or maybe  we could do better?

**Review by xuchenhan-tri — COMMENTED** [review]
> Reviewed 1 of 1 files at r5.
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/16243)* status: 1 unresolved discussion, needs platform reviewer assigned, needs at least two assigned reviewers (waiting on @amcastro-tri)

___
*[multibody/contact_solvers/sap_solver.h, line 26 at r2](https://reviewable.io/reviews/robotlocomotion/drake/16243#-Mr8njfY9dsi7UjCr_S7:-Mr9iNa_4CTAnwio_Xev:b6hdt6a) ([raw file](https://github.com/robotlocomotion/drake/blob/4421e18e6c2f6ae345f8d86b95140c3cb705de9d/multibody/contact_solvers/sap_solver.h#L26)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

I added more docs. Hopefully that clarifies? or maybe  we could do better?
</blockquote></details>

That's helpful. Thanks!
___
*[multibody/contact_solvers/sap_solver.h, line 38 at r5](https://reviewable.io/reviews/robotlocomotion/drake/16243#-Mr9i3l-042gzTNiwrIx:-Mr9i3l-042gzTNiwrIy:brpnc06) ([raw file](https://github.com/robotlocomotion/drake/blob/3ef551afd63f6852e3fdc896d506efafe3e9e872/multibody/contact_solvers/sap_solver.h#L38)):*
> ```objc
>   // errors.
> 
>   // Stopping criteria tolerances. We monitor the optimality condition (for
> ```

BTW, consider rephrase for better flow



_Suggestion:_
```objc
  // Optimality condition criterion: We monitor the optimality condition (for
```
___
*[multibody/contact_solvers/sap_solver.h, line 61 at r5](https://reviewable.io/reviews/robotlocomotion/drake/16243#-Mr9iUH631HyC-tTg8ni:-Mr9iUH7EUJ_3agq-SxQ:b-4xd41m) ([raw file](https://github.com/robotlocomotion/drake/blob/3ef551afd63f6852e3fdc896d506efafe3e9e872/multibody/contact_solvers/sap_solver.h#L61)):*
> ```objc
>   double rel_tolerance{1.e-6};   // Relative tolerance εᵣ.
> 
>   // We monitor the decrease of the cost on each iteration. It is not worth it
> ```

BTW, consider rephrase for better flow



_Suggestion:_
```objc
  // Cost condition criterion: We monitor the decrease of the cost on each iteration. It is not worth it
```

**Review by amcastro-tri — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/16243)* status: 1 unresolved discussion, needs platform reviewer assigned, needs at least two assigned reviewers (waiting on @xuchenhan-tri)

___
*[multibody/contact_solvers/test/sap_solver_test.cc, line 454 at r2](https://reviewable.io/reviews/robotlocomotion/drake/16243#-Mr907rJ1-HyVbJ_eKNF:-Mr9joOj69riOZSSDILv:b-53wb0q) ([raw file](https://github.com/robotlocomotion/drake/blob/4421e18e6c2f6ae345f8d86b95140c3cb705de9d/multibody/contact_solvers/test/sap_solver_test.cc#L454)):*
<details><summary><i>Previously, xuchenhan-tri wrote…</i></summary><blockquote>

Now that we have two exit conditions and test that insists on a particular exit condition being triggered, `v_guess` being nonzero is not just an extra mile we go to stress the solver, but a necessity for the test to pass. 

In fact, I just tried changing `v_guess` to zero and the unit test failed.
</blockquote></details>

Good catch, I'm nor surprised since v_guess=0 is a great guess near the steady state. Added note.

**Review by xuchenhan-tri — COMMENTED** [review]
> Reviewed all commit messages.
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/16243)* status: needs platform reviewer assigned, needs at least two assigned reviewers (waiting on @sherm1)

**Review by xuchenhan-tri — COMMENTED** [review]
> Reviewed 1 of 1 files at r6.
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/16243)* status: needs platform reviewer assigned, needs at least two assigned reviewers (waiting on @amcastro-tri)

**Review by xuchenhan-tri — COMMENTED** [review]
> Reviewed 1 of 1 files at r7, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/16243)* status: needs platform reviewer assigned, needs at least two assigned reviewers (waiting on @amcastro-tri)

**Review by amcastro-tri — COMMENTED** [review]
@sherm1, you were not assigned. Fixing now.
+@sherm1 

> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/16243#-:-Mr9rbqT7RgSoslyL0Wj:bchj34g)* status: :shipit: complete! all discussions resolved, LGTM from assignees sherm1(platform),xuchenhan-tri (waiting on @amcastro-tri)
